### PR TITLE
Use latest haskell-language-server, remove custom package

### PR DESCRIPTION
**Pull request by mcwitt**
Also removes custom derivation, which shouldn't be necessary anymore since recent HLS versions are in Hackage.

### CODE REVIEW

**Comment by mcwitt** [comment]
This depends on https://github.com/hasktorch/hasktorch/pull/532 to add an upper bound for `singletons` (otherwise the Hackage upgrade breaks the build).

**Comment by tscholak** [comment]
Woohoo, thanks a lot!

**Comment by mcwitt** [comment]
I see we're now using `haskell-language-server = "latest"` in the main hasktorch [repo](https://github.com/hasktorch/hasktorch/blob/63b3df06477ab1fc60030ccbb630330e71fdca74/shell.nix#L31). I'll change it to "latest" here too once https://github.com/hasktorch/hasktorch/pull/532 is merged.

**Comment by tscholak** [comment]
can I ask you to also bump the rev's in nix/sources.json in the main hasktorch repo to bring these into alignment? When I updated to hls 1.0 in the main repo, hls 1.0 wasn't on hackage yet and I had to use the custom tools code. I think you can remove it there, too. I don't have a preference with regards to "latest" or "1.0.0". :)

**Comment by mcwitt** [comment]
> can I ask you to also bump the rev's in nix/sources.json in the main hasktorch repo to bring these into alignment?

No problem, did this in https://github.com/hasktorch/hasktorch/pull/533.

I tried bumping all the dependencies with `niv update` there (for this PR I only updated `hackage.nix`). If it goes well with upgrading all dependencies, seems like we should do that here too?

**Comment by tscholak** [comment]
I had some problems with nixpkgs in the past. In general, it seems best to use exactly what the haskell.nix folks are using, cf. https://github.com/input-output-hk/haskell.nix/blob/master/nix/sources.json#L107

**Comment by mcwitt** [comment]
Ah, it seems like we're automatically using the haskell.nix nixpkgs in hasktorch because `nixpkgs-staging-next` doesn't exist in the niv sources: https://github.com/mcwitt/hasktorch/blob/129be3f2ae23e20521859baa70e739066e8db490/nix/default.nix#L26-L27

But in this repo, it seems we are overriding haskell.nix's version with the `nixpkgs-staging-next` in the sources.

I wonder if we shouldn't just use `iohKNix.nixpkgs` here too?

**Comment by tscholak** [comment]
hmm, I don't know about `iohkNix.nixpkgs`. Perhaps we should use: https://github.com/input-output-hk/haskell.nix/blob/master/default.nix#L27. There's a tiny chance that we will not have to bootstrap ghc anymore if we do that.

**Comment by mcwitt** [comment]
> hmm, I don't know about iohkNix.nixpkgs. Perhaps we should use: https://github.com/input-output-hk/haskell.nix/blob/master/default.nix#L27. There's a tiny chance that we will not have to bootstrap ghc anymore if we do that.

@tscholak something like 8ce772f seems to be the current [recommended](https://input-output-hk.github.io/haskell.nix/tutorials/getting-started/#scaffolding) way. I might be missing some subtlety, though; does this seem reasonable to you?

If I'm interpreting correctly, we also have several unused nixpkgs versions in `sources.json` (removed in 841778f)

File: hasktorch-skeleton.cabal [hunk 1]
```diff
@@ -1,4 +1,4 @@
-cabal-version:       2.1
+cabal-version:       1.24
```
Comment (mcwitt) [inline hasktorch-skeleton.cabal:1]: This is the same `cabal-version` used by hasktorch. When I upgrade to GHC 8.10.4 without changing this, I get

```
hasktorch-skeleton.cabal:0:0: error:
Unsupported cabal-version 2.1. See https://github.com/haskell/cabal/issues/4899.
```

**Comment by mcwitt** [comment]
@tscholak looks like we are running out of disk space on the CI runners :confused: 

https://github.com/hasktorch/hasktorch-skeleton/runs/2131770949

**Comment by junjihashimoto** [comment]
It seems that CI builds previous ghc versions like ghc-8.8.4, so this CI takes some hours.
I still don't know what the cause is.

File: nix/default.nix [hunk 2]
```diff
@@ -19,55 +19,29 @@ assert cudaSupport -> (cudaMajorVersion == "9" || cudaMajorVersion == "10" || cu
...
-  haskellNix = (import sources.haskell-nix { inherit system sourcesOverride; }).nixpkgsArgs;
-  # use our own nixpkgs if it exist in our sources,
-  # otherwise use iohkNix default nixpkgs.
-  nixpkgs = sources.nixpkgs-staging-next or
-    (builtins.trace "Using IOHK default nixpkgs" iohKNix.nixpkgs);
+  iohkNix = import sources.iohk-nix {};
+  haskellNix = import sources.haskell-nix { inherit system sourcesOverride; };
+
+  # Use haskell.nix default nixpkgs
+  nixpkgsSrc = haskellNix.sources.nixpkgs-2009;
```
Comment (tscholak) [inline nix/default.nix]:
```suggestion
  nixpkgsSrc = haskellNix.sources.nixpkgs-unstable;
```
cf. https://github.com/hasktorch/hasktorch/blob/master/nix/default.nix#L26.
This should help with the build.

Comment (junjihashimoto) [reply nix/default.nix]: Thx! I'll try it.

Comment (junjihashimoto) [reply nix/default.nix]: https://github.com/junjihashimoto/hasktorch-skeleton/actions
There is no reduction for nix-build. In the stack's case, it is 1 hour shorter.

image

Comment (mcwitt) [reply nix/default.nix]: @junjihashimoto I hope that b00d914 will help; this brings the dependency versions into sync with the hasktorch repo so we should get more cache hits.

Comment (mcwitt) [reply nix/default.nix]: Also, I had neglected to update the hasktorch dependency and GHC version in `stack.yaml`. Fixed that in 8c0a35e and a12460e.

**Review by tscholak — APPROVED** [review]
The cache is hit as expected. Well done!

Looks all good to me, thank you very much!

**Comment by junjihashimoto** [comment]
You’ve been very helpful.

**Pull request merged**